### ISSUE TITLE
[Suggestion] Admin log in as customer

### ISSUE THREAD

**Issue by judygunderson** [zencart/zencart#2680]
This is a mod that is requested A LOT by my customers and I think would be great to be included in future versions of Zen Cart.

**Comment by wiztechinc** [comment]
Since you can now change the customer's password, it's not as necessary but so many issues my clients ask about could be taken care of with this mod. Sure does give a cart owner control when buyers are clueless about  - well, just about everything it seems like. However, I don't install it very often cause I'm all about keeping the number of mods down and we figure out other ways to deal with issues. Cart owners might really like this as a default part of zen cart.

**Comment by scottcwilson** [comment]
Definitely a step saver over using Master Password and logging in that way.

**Comment by torvista** [comment]
Regularly, the crap bank payment gateway I need to use does not complete the order post-payment so I need to do this to complete the order with another payment option, just to record it. 
So, yes I need this feature.

A great deal of discussion and effort could be saved by just importing lat9's entire repository!
There is little/nothing there that is not of use and the code quality is core-quality+.

**Comment by judygunderson** [comment]
I am guessing my customers use it to take credit card phone orders - you know how customers can be - too lazy to do it themselves I guess!  giggle  
Delia - I'm with you on keeping the mods down.  I am constantly discouraging customers from loading their carts with mods they may never or seldom ever use!

**Comment by lat9** [comment]
> 
> 
> Definitely a step saver over using Master Password and logging in that way.

If you're referring to ]Encrypted Master Password](https://github.com/lat9/encrypted_master_password), I don't understand the step-saving.

**Comment by scottcwilson** [comment]
You don't have to type the customer's email or the password.  Just press the place order button.  (It's super minor I know but these little things add up for a busy shop that's processing a lot or orders.)

**Comment by lat9** [comment]
Oh, the functionality that was present in EMP until DrByte requested its removal?

**Comment by scottcwilson** [comment]
Not sure of the history - it might have had the same security issue that ALAC used to have.  At any rate, let's stay on ALAC for this thread please.

**Comment by drbyte** [comment]
I haven't looked at ALAC in awhile, but last time I did I was concerned that it created security issues.
Plus IIRC it was very invasive code (ie: changed a ton of core file logic for the little value it offered).
Maybe it's improved since then.

**Comment by scottcwilson** [comment]
It has!  The security issue was fixed.  https://www.zen-cart.com/downloads.php?do=file&id=583

**Comment by lat9** [comment]
> Not sure of the history - it might have had the same security issue that ALAC used to have. At any rate, let's stay on ALAC for this thread please.
Don't know.  At the time @drbyte had sent me a PM indicating that he considered the auto-login feature as having the potential to "give away the keys to the kingdom".

**Comment by barco57** [comment]
ALAC requires the installation of Master Password according to the note on the module

> NOTE: This module relies on the Master Password contribution, which MUST BE INSTALLED SEPARATELY

What I didn't like about Master Password was the was a singular Master password so employees had to know there admin login password and maintain control over a Master password as well. I have always preferred Encrypted Master Password and it doesn't use a singular master password. Click the login button from the selected customer and you are brought to the front end log in page with the email address already filled out and you enter your specific admin password instead of a master password. I would prefer that we are tracking which admin placed the order for the customer which is more easily tracked by the use of their assigned admin password I would think... not a coder disclaimer here...blah, blah, blah stupid normy

**Comment by scottcwilson** [comment]
We don’t have to use the same approach as the mod uses. :)

**Comment by dbltoe** [comment]
Not sure if any changes have come down the pipe since last I tested the two together but, EMP made it easier to manage all a caller's account info as well as make an order.  EMP also generally showed items already in the cart if the user was having difficulty.  Of course the ability to restrict staff from doing damage elsewhere by setting a profile is a definite plus.  Best of all, is the lat9 part.

**Comment by scottcwilson** [comment]
> A great deal of discussion and effort could be saved by just importing lat9's entire repository!

Like Michael Bolton, I celebrate her entire catalog! :)

**Comment by judygunderson** [comment]
Scott, I could be wrong, but I think you just aged yourself with the Michael Bolton (God love him and his music) comment!  lol

**Comment by scottcwilson** [comment]
It's a joke from Office Space, the greatest movie of all time. 

https://www.amazon.com/Office-Space-Special-Screen-Century/dp/B01GWCLH0Q/ref=sr_1_5

**Comment by lankeeyankee** [comment]
"EMP made it easier to manage all a caller's account info as well as make an order. EMP also generally showed items already in the cart if the user was having difficulty"

This says it all. We use this daily, it's essential. The benefits are huge when you are on the phone with a customer. We just ask them for their email address and type it in. It's not really a biggie, and you'd have to do the same to find them in the admin to click the ALAC button. So it's not any extra time really.

**Comment by scottcwilson** [comment]
The current approach in ALAC is predicated on having a plain text MASTER_PASS available.  If we don't do this, we'd need to prompt for an admin password to place the order (you need plain and encrypted to validate).  This is more secure but also more annoying ... still noodling on possible approaches; interested in input from developers who have thought about this.

**Comment by wiztechinc** [comment]
I do not understand why there needs to be a master password now. Once the password is in, anyone who is in admin can do whatever. Different admin groups now allow a site owner to restrict others as to where they can go. I can't come up with a reason that makes a password necessary. Perhaps we need to look in this light as well?

**Comment by scottcwilson** [comment]
The password is required to authenticate the request on the catalog side.  Take a look at the code and run through some scenarios and you'll see why.

**Comment by wiztechinc** [comment]
who me? dig into code... hmmm....

**Comment by judygunderson** [comment]
Delia - you made me giggle with that statement because that is exactly what I would have said!  LOL

**Comment by wiztechinc** [comment]
Judy, we may not be code rats but they sure need us, if just for the laughs!

**Comment by judygunderson** [comment]
As long as they don't ask us to sing and dance, I'm good with just relieving their tensions with a few laughs now and then Delia!  giggle
Alternatively they could tell us to shut the heck up so they can concentrate...... YIKES!!

**Comment by dbltoe** [comment]
Sorry, just can't see ALAC over EMP.  EMP already set for encrypted and settings allow defining just who can use it.

**Comment by barco57** [comment]
I suppose I will ask, what does ALAC bring to the table that EMP doesn't? As far as I can tell they do the same thing and EMP does it much more elegantly, meaning no overrides of core files, no template files,

**Comment by zcwilt** [comment]
First of all, I have only a cursory knowledge of both ALAC and EMP although more for ALAC as I have upgraded sites that use it.

My question is mainly for those that use EMP. 
In what way is data encrypted. From a cursory look at the code there isn't any encryption involved, it merely posts the admin password directly to the login page and then uses an observer to validate that password.

**Comment by scottcwilson** [comment]
For some reason I thought this feature had been removed.  The way it works looks perfectly fine to me, and I agree with @barco57 and @dbltoe  that this mod is a better match for the desired functionality. 

SO!  After all that - any objections from @zcwilt or @drbyte for me building a PR that includes the EMP Logic?  (Unless @lat9 wants to do it?)

**Comment by scottcwilson** [comment]
> Ok my bad, it doesn't post to the login page but admin can login with their password. Still no encryption though.

None is required since it does a password prompt.  This means it's secure in a way that ALAC is not.

**Comment by zcwilt** [comment]
Maybe the plugin should change it's name then. ;)

Given that, I do like that EMP is logging stuff. 

There will be those that don't like the fact that there is that extra step, e.g. having to login over ALAC's automated process, but people could still install ALAC if they wanted to.

Given that I'm OK with having a PR.

**Comment by lat9** [comment]
@scottcwilson, I'll be happy to do the EMP integration PR.  How about I provide a minor update (and configuration setting) to restore the handling to allow an automatic login?

> 
> 
> Maybe the plugin should change it's name then. ;)
> 
I'd love to, but that's been its name since I took over support.  It's kind of like changing your moniker from @zcwilt!

**Comment by scottcwilson** [comment]
Can you just briefly describe your design for a no-login approach?  I can't think of a great way to do it securely, but there are certainly solutions that involve opening up a brief window, and I want to be sure the core team is OK with whatever you choose.

**Comment by zcwilt** [comment]
@scottcwilson @lat9 
For automated login, I would look to getting rid of the Master Password. It's probably the weak link, in that unless we enforce some parameters on it, there's a good chance it will be very weak, even then it's going to be a static value which people will rarely change, which in itself is a weakness.

As ALAC already stuffs an encrypted password into the db per customer, then a better way of doing this  is to use a random salt/hash again stuffed in the db. The link to the customer side would then contain an hmac encoded value using time, customer email customer id etc. 
Catalog side could then use hmac verification of the value.

**Comment by zcwilt** [comment]
I guess even if master password was originally autogenerated that would be a step forward.
That's pretty much how shopify allow verification for apps.

**Comment by drbyte** [comment]
Can there just be temporary session substitution, where the admin can impersonate any customer and the session gets filled up as though it's the customer but with an extra admin-impersonation value and then when the admin is done they end the impersonation and it swaps the customer session stuff out and puts the admin session back?

**Comment by lat9** [comment]
There is **no** `Master Password` setting; the processing uses an admin's current admin-password for verification.  From EMP v1.8.0 (please note that the code's based on an oldy-moldy version of Zen Cart), the `Place Order` button addition looks like:
```
//-bof-a-EMP login link 1/1
        $emp_sql = 'SELECT admin_profile, admin_pass from ' . TABLE_ADMIN . ' WHERE admin_id = :adminId:'; /*v1.7.0c*/
        $emp_sql = $db->bindVars($emp_sql, ':adminId:', $_SESSION['admin_id'], 'integer');
        $emp_result = $db->Execute($emp_sql);

        if (!$emp_result->EOF && ($_SESSION['admin_id'] == EMP_LOGIN_ADMIN_ID || $emp_result->fields['admin_profile'] == EMP_LOGIN_ADMIN_PROFILE_ID) ) {  /*v1.7.0c*/
          $hash = md5($cInfo->customers_email_address . $cInfo->customers_id . $_SESSION['admin_id'] . $emp_result->fields['admin_pass']);  /*v1.7.0a*/
          $contents[] = array('align' => 'center', 'text' => '<div align="center"><form target="_blank" name="login" action="' . zen_catalog_href_link(FILENAME_LOGIN, 'action=process', 'SSL') . '" method="post">' . zen_draw_hidden_field('email_address', $cInfo->customers_email_address) /*-bof-a-v1.7.0*/ . zen_draw_hidden_field('cID', $cInfo->customers_id) . zen_draw_hidden_field('aID', $_SESSION['admin_id']) . zen_draw_hidden_field('keyValue', $hash) /*-eof-a-v1.7.0*/ . zen_image_submit('button_placeorder.gif', EMP_BUTTON_PLACEORDER_ALT) . '</form></div>');
          
        }
//-eof-a-EMP login link 1/1
```

This gets intercepted by a storefront init_include which sets the `$emp_login` variable for use in the login-page's header:

```
/*-----
** If the current page is the login page and it's a login request from the admin's Customers->Customers page,
** fix up the securityToken.
** Note:  Totally reworked for v1.6.0!
*/
if (isset($_GET['main_page']) && $_GET['main_page'] == FILENAME_LOGIN && strpos($_SERVER['HTTP_REFERER'], '/customers.php') !== false) {
  $emp_login = false;
  if (isset($_POST['email_address']) && isset($_POST['cID']) && isset($_POST['aID']) && isset($_POST['keyValue'])) {
    $check_emp_query = "SELECT customers_id FROM " . TABLE_CUSTOMERS . " WHERE customers_email_address = :emailAddress AND customers_id = :customersID";
    $check_emp_query = $db->bindVars($check_emp_query, ':emailAddress', zen_db_prepare_input($_POST['email_address']), 'string');
    $check_emp_query = $db->bindVars($check_emp_query, ':customersID', (int)$_POST['cID'], 'integer');
    $check_emp = $db->Execute($check_emp_query);
    if (!$check_emp->EOF) {
      $check_emp_query = 'SELECT admin_pass FROM ' . TABLE_ADMIN . ' WHERE admin_id = :adminID';
      $check_emp_query = $db->bindVars($check_emp_query, ':adminID', (int)$_POST['aID'], 'integer');
      $check_emp = $db->Execute($check_emp_query);
      if (!$check_emp->EOF) {
        if ($_POST['keyValue'] === md5($_POST['email_address'] . $_POST['cID'] . $_POST['aID'] . $check_emp->fields['admin_pass'])) {
          $_POST['securityToken'] = $_SESSION['securityToken'];
          $emp_login = true;
        }
      }
    }
  }
}
```

In that login-page header, the auto-login takes precedence but still provides configured admins to provide _their_ admin password instead of the customers:
```

    if (!$check_customer->RecordCount()) {
      $error = true;
      $messageStack->add('login', TEXT_LOGIN_ERROR);
    } elseif ($check_customer->fields['customers_authorization'] == '4') {
      // this account is banned
      $zco_notifier->notify('NOTIFY_LOGIN_BANNED');
      $messageStack->add('login', TEXT_LOGIN_BANNED);
      
      // Check that password is good
//-bof-Encrypted Master Password by stagebrace *** 1/1
      //- Start modifications by lat9:
//-bof-a-v1.6.0
    } elseif ($emp_login === true) {
      $ProceedToLogin = true;
      $_SESSION['emp_admin_login'] = true;
      $_SESSION['emp_admin_id'] = (int)$_POST['aID'];
//-bof-a-v1.7.0
      $sql_data_array = array( 'access_date' => 'now()',
                               'admin_id' => $_SESSION['emp_admin_id'],
                               'page_accessed' => 'login.php',
                               'page_parameters' => '',
                               'ip_address' => substr($_SERVER['REMOTE_ADDR'],0,45),
                               'gzpost' => gzdeflate( json_encode( array ( 'action' => 'emp_admin_login' )), 7),
                               'flagged' => 0,
                               'attention' => '',
                               );
      zen_db_perform(TABLE_ADMIN_ACTIVITY_LOG, $sql_data_array);
//-eof-a-v1.7.0
//-eof-a-v1.6.0
      
    } else {
      //--- Either specific admin ID or an admin within a specific admin profile can use their password to login.
      if (!defined('EMP_LOGIN_ADMIN_ID')) define ('EMP_LOGIN_ADMIN_ID', 1);  /*lat9-a/c*/
      $get_admin_query = "SELECT admin_id, admin_pass
                          FROM " . TABLE_ADMIN . "
                          WHERE admin_id = " . (int)EMP_LOGIN_ADMIN_ID;  /*lat9-c*/
      $check_administrator = $db->Execute($get_admin_query);
      $customer = (zen_validate_password($password, $check_customer->fields['customers_password']));
      $administrator = (zen_validate_password($password, $check_administrator->fields['admin_pass']));
      if ($customer) {
        $ProceedToLogin = true;
        
      } elseif ($administrator) {
          $ProceedToLogin = true;
          $_SESSION['emp_admin_login'] = true;  /*lat9-a*/
          $_SESSION['emp_admin_id'] = EMP_LOGIN_ADMIN_ID;
          
      } else {
//-bof-lat9-a
        if (!defined('EMP_LOGIN_ADMIN_PROFILE_ID')) define ('EMP_LOGIN_ADMIN_PROFILE_ID', 1);  /*lat9-a*/
        $admin_profile_query = "SELECT admin_id, admin_pass 
                                  FROM " . TABLE_ADMIN . " 
                                  WHERE admin_profile = " . (int)EMP_LOGIN_ADMIN_PROFILE_ID;
        $admin_profiles = $db->Execute($admin_profile_query);
//-eof-lat9-a
        $ProceedToLogin = false;
//-bof-lat9-a
        while (!$admin_profiles->EOF && !$ProceedToLogin) {
          $ProceedToLogin = zen_validate_password($password, $admin_profiles->fields['admin_pass']);
          if ($ProceedToLogin) {
            $_SESSION['emp_admin_login'] = true;
            $_SESSION['emp_admin_id'] = $admin_profiles->fields['admin_id'];
          }
          $admin_profiles->MoveNext();
        }
//-eof-lat9-a
      }
    }
      if (!($ProceedToLogin)) {
//-eof-Encrypted Master Password by stagebrace *** 1/1
```

**Comment by zcwilt** [comment]
@lat9 
Sorry, I was referring to how ALAC does auto login, having only looked at the most recent version of EMP hadn't seen old code that used auto login.

**Comment by zcwilt** [comment]
In terms of security though, I still see the benefit of using a rotating/random hmac key.

**Comment by scottcwilson** [comment]
Much better than a pass in clear text.

**Comment by zcwilt** [comment]
However at least to move things forward, lets get an initial PR going based on current code.

**Comment by lat9** [comment]
Will do, @zcwilt.

**Comment by lat9** [comment]
Question:  Should I provide a _full_ integration or continue to use the EMP observer-class?

**Comment by barco57** [comment]
I come back from the dentist to a flurry of activity, did I start that?

**Comment by zcwilt** [comment]
@lat9 
Full integration.

**Comment by lat9** [comment]
OK; I'm _assuming_ that the feature should be renamed for incorporation into the core.  There are currently two configuration settings (in `My Store`):

1. Encrypted Master Password: Single Admin ID
2. Encrypted Master Password: Admin Profile ID

I'm thinking to rename these settings, respectively, to

1.  Customer Order Placement: Single Admin ID
2.  Customer Order Placement: Admin Profiles List

**Comment by scottcwilson** [comment]
I'd rather see you create a page and handle it like any other page, where superuser always has access and other users have access on a per profile basis.

**Comment by lat9** [comment]
Some of my clients use the single-admin-id to restrict the feature, even for superusers.  Remember, this feature, from the admin, simply adds a `Place Order` button iff the admin qualifies based on these settings.

**Comment by scottcwilson** [comment]
What if we were to add this as a per admin user configuration parm, say on Admins->Admin Users?  Might possibly be useful for other sensitive data?  Just a suggestion.

**Comment by lat9** [comment]
@zcwilt's guidance was to start with the plugin as-is.  Your suggestion is solid, but let's get the basics integrated first.

**Comment by drbyte** [comment]
Question: What's the goal here? 

There are completely different meanings to these phrases:
"Admin Login As Customer" 
"Encrypted Master Password"
"Customer Order Placement"
"Place Order As Customer"
"Login To Customer Session"

I don't think any mention of "Encrypted Master Password" has any relevant meaning at all -- it's all just technical geek-speak as far as a storeowner is concerned.

Is the primary purpose just to "login as the customer in order to place an order 'as them'" (Or in the case of a payment gateway failure, "finish the order", which IMO is largely the same purpose.) 

If the primary goal is to "Place Order As Customer", then let's push all the terminology in that direction.

Granted, an even better, but more involved, solution would be an admin UI for placing orders.

**Comment by dbltoe** [comment]
Doesn't matter if it says "Rover, Rover, let Freddie come over " or any other text.  Just feel that the code should be that of EMP.  I tell customers to use it as if it were a Log In as Customer.
Even the original author touted it as "allows the stores administrator to login to any account with their administrator password."  Probably should have been better named then.

**Comment by judygunderson** [comment]
I personally like it because when a customer calls I can pull them up by their last name, then click the "place order" button and go.  Easy peasy.
Many people don't even know what email address they used to create an account... so you have to look up the customer anyway to get their email address to then copy, then paste using EMP.  I just think ALAC is much smoother and easier... IMHO. :)

**Comment by barco57** [comment]
@judygunderson your thinking of the original master password mod... EMP has had the place order button for along time possibly since it was created, takes you to the front end with the email filled out, enter your admin password and your in...

**Comment by zcwilt** [comment]
Have to agree with @drbyte that the terminology here needs clarifying.
As we are looking to integrate EMP, what does it actually do.
basically it just lets an admin login as a customer. 
So text suggesting it just allows you to `Place an order` is misleading, even though that might be the main usage.

Again, Chris has also suggested that having Admin code to allow creation/changing of orders is probably better, don't disagree, but that may be some way off.
In fact there is a lot more admin side stuff that could be done in managing customer information
e.g. adding/editing address book entries 
that would eventually make this redundant. 
But we aren't close to that, so this seems a fair first step.

**Comment by torvista** [comment]
I think it should be Login as Customer: what you do after is moot.

Do it in the Admin?....you are talking about viewing a cart-in-progress and making an action to convert it to an order...maybe, but sometimes I have to pick a product for a customer. This is going down a road of duplicating catalog code, so why bother.

One point I would make which gets in the use of this mod is in my case LastPass (probably other ps mangers do the same) autofilling/overwriting the field on the login page which has already been auto-filled by EMP. I would much prefer that a click on "Login as Customer" takes me straight to a logged-in cart-view or if no cart, the account page, bypassing the manual login, if such as thing is possible.
Remember this process is usually done with one hand, the other is on the phone, so "an extra couple of clicks" or a copy-paste is significant in this case.

**Comment by judygunderson** [comment]
Barco57 - I didn't know that because I have not used it since ALAC came out, so thanks for the info.  

However, Like Steve said - I also have LastPass - and I am guessing it would override anything from EMP with my admin login or something - thus making it double work for me.   having to go back in and get their email address and the master password to enter.  Whereas ALAC does not have that issue of having to fill in a form.

_Originally posted by @barco57 in https://github.com/zencart/zencart/issues/2680#issuecomment-518499573_

**Comment by gburton** [comment]
Put a "Key" against each customer, eg:  abcde-12345-fghij-67890 (obviously way more random than that for each).  

In admin, on customers_info page, have a link to shopside/go/{{key}}

Click the link...shopside opens in a new page with Admin (as Customer) logged in.  

You could extend that further by allowing Customers to login using the Key, now you have the making of a passwordless login system.

**Comment by dbltoe** [comment]
We need to be mindful of what an owner will be willing to grant as far as permissions for someone logging in as the customer.  The admin profiles can take care of this by either having it a separate option or make sure the link is in an area available to the one taking the order.
I can see a scenario where some may have the authority to discount an order but others not.  Likewise, some allowed to edit and some just to take an order.
Only drawback is to attain this without winding up with 47 admin profiles.
That's why I find the login page the most convenient.  Anyone can get to that page and perform taking an order.  No need to even give them access to the admin order or customer area.  Leave higher level to the up line.
As it sits now, the comma-separated list allows the low- and high-level employees to do their thing.

**Comment by lat9** [comment]
FWIW, I've just resurrected the auto-login feature for the EMP plugin: https://github.com/lat9/encrypted_master_password/commit/600f076881c21cbc66b7cc9a42743ab0f1458a25

**Comment by zcwilt** [comment]
I've been working on code to do EMP autologin as well. Code is almost complete and should have a PR tomorrow.

Code works as follows :-

Each store will have a GLOBAL_AUTH_KEY, this is a sha256 hash of a cryptographically random number. 

Each customer has customers_secret field, this again is a  sha256 hash of a cryptographically random number however it is updated on each use of the auto login, in effect creating a nonce.

Auto login will only work if store is using ssl, obviously to mitigate leakage of post data.

An hmac key (which is the only data passed in the url) is created from post data, which consists of 
 - customer id
 - admin id 
 - customer email address
 - timestamp

with a concatenation of the GLOBAL_AUTH_KEY and customers_secret being used as the hmac creation key.

The timestamp is used store side to limit the window of opportunity.

**Comment by lat9** [comment]
@zcwilt, some sites have a **_large_** number of customers and I'm wondering how much overhead is involved in the addition of that 'secret' field to each and every customer.  Is there a reason that the secret-value is customer- rather than admin-dependent?

**Comment by zcwilt** [comment]
@lat9
There's very little overhead. The customer_secret is only ever set/updated on the admin customers page when the `Place Order` button is available.

The reason for using it over just a global admin key is to provide some extra randomness to the hmac key.

**Comment by lat9** [comment]
Thanks for the confirmation, @zcwilt.  I await your update.

**Comment by scottcwilson** [comment]
Can this issue be closed now that 1.5.7 has a Place Order button on the Customers screen?

**Comment by scottcwilson** [comment]
This change seems to be done.  Can we close?
